File: 094-inorder-traversal/Solution.java
import java.util.ArrayList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) {
        this.val = val;
    }
}
class Solution94 {
    List<Integer> list = new ArrayList<>();
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null)
            return list;
        inorderTraversal(root.left);
        list.add(root.val);
        inorderTraversal(root.right);
        return list;
    }
}
/*
输入：root = [1,null,2,3]
输出：[1,3,2]

遞迴樹:
inorderTraversal(root=1)
│
├─ inorderTraversal(root.left=null)
│  └─ (返回，无操作)
│
├─ list.add(1)  →  list = [1]
│
└─ inorderTraversal(root.right=2)
   │
   ├─ inorderTraversal(2.left=3)
   │  │
   │  ├─ inorderTraversal(3.left=null)
   │  │  └─ (返回，无操作)
   │  │
   │  ├─ list.add(3)  →  list = [1, 3]
   │  │
   │  └─ inorderTraversal(3.right=null)
   │     └─ (返回，无操作)
   │
   ├─ list.add(2)  →  list = [1, 3, 2]
   │
   └─ inorderTraversal(2.right=null)
      └─ (返回，无操作)
      
總結精隨:
可以理解成, 一直往左邊走, 走到不能再走的時候才開始做list.add()
*/

File: 094-inorder-traversal/solution.go
/*
中序遍歷, 迭代版
*/
package main

import (
	"fmt"
	"gopractice/datastructure/Tree"
)

type TreeNode = Tree.TreeNode

func inorderTraversal(root *TreeNode) (result []int) {
	stack := make([]*TreeNode, 0)
	cur := root
	for cur != nil || len(stack) > 0 {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		//出棧 + 塞result
		n := len(stack)
		cur = stack[n-1]
		stack = stack[:n-1]
		result = append(result, cur.Val.(int))

		cur = cur.Right
		fmt.Printf("stack = %v\n result = %v,", stack, result)
	}
	return
}
func main() {
	array := []any{1, 2, nil, 3, 5, nil, nil, 4}
	root := Tree.BuildLevelOrderTree_v2(array)
	inorderTraversal(root)
}

/*
左根右

stack = [1], result = []
stack = [], result = [1]
stack = [2], result = [1]
stack = [2, 3], result = [1]
stack = [2], result = [1, 3]
stack = [], result = [1, 3, 2]

用標準的迭代中序（先往左一路推入 stack，然後逐個 pop 並處理，最後移到右子樹）。
*/
